### PR TITLE
Fix warning: Undefined index: instance_id

### PR DESCRIPTION
**Pull request by ThaDafinser**
WARNING: ...\piwik\core\SettingsPiwik.php(408): Notice - Undefined index: instance_id - Piwik 2.14.1 - Please report this message in the Piwik forums: http://forum.piwik.org (please do a search first as it might have been reported already)

### CODE REVIEW

**Comment by mnapoli** [comment]
Weird, the `@` should prevent the warning from appearing (even with our custom error handler: https://github.com/piwik/piwik/blob/master/core/ErrorHandler.php#L69-L72). I don't understand why the warning appeared…

File: core/SettingsPiwik.php [hunk 1]
```diff
@@ -405,9 +405,9 @@ protected static function getPiwikInstanceId()
             return false;
         }
 
-        $instanceId = @Config::getInstance()->General['instance_id'];
-        if (!empty($instanceId)) {
-            return $instanceId;
+        $configGeneral = Config::getInstance()->General;
+        if(isset($configGeneral['instance_id']) && !empty($configGeneral['instance_id'])){
```
Comment (sgiehl) [inline core/SettingsPiwik.php]: There is no need to check ```!isset``` when using ```!empty```.

**Comment by ThaDafinser** [comment]
@mnapoli not with `xdebug.scream = On` setting
IMO we should try to get rid of the @

**Comment by mnapoli** [comment]
Ah right, then I would say it's a wontfix then. We use the error suppressor everywhere in Piwik. I'm not saying it's good, but removing it here because of xdebug (which shouldn't be installed or enabled in production) wouldn't make sense if we leave it everywhere else.

**Comment by ThaDafinser** [comment]
@mnapoli i was developing a plugin and wanted to test it in development, therefor i need the Piwik installation.

But i know what thats only one place of many.

**Comment by sgiehl** [comment]
Shouldn't we work towards removing all of the error suppression? So why not starting with this one?

**Comment by ThaDafinser** [comment]
To remove all suppressions from the config it would be easiert to change this `Config::getInstance()->General['instance_id']` to something like  ` Config::getInstance()->get('general', 'instance_id')`

**Comment by mnapoli** [comment]
> i was developing a plugin and wanted to test it in development, therefor i need the Piwik installation. (I dont want to change my php.ini settings because of one project)

Makes sense. However the "scream" option is meant to achieve exactly what you get here, and currently `@` is accepted practice so it's kind of incompatible (or at least, inconvenient for you).

So we get back to this point:

> Shouldn't we work towards removing all of the error suppression? So why not starting with this one?

I wouldn't be against discussing that with pros and cons (and I'm not against removing `@`), but I'm just trying to go in the correct order of things :) The error suppressor is not deemed "bad" in current Piwik practices and it's still used and introduced in new code. So unless we decide to change that, it doesn't make sense to start removing it (since it will be added again later in other places).

TL/DR: before removing `@` we should discuss and decide that we shouldn't use anymore?

**Comment by mattab** [comment]
> TL/DR: before removing @ we should discuss and decide that we shouldn't use anymore?

:+1: for now, we don't need to remove the extras `@` in front of INI config setting reading. (as said, it's regularly done in the codebase, and even in some of the new code)
